frex/utils/constraint_solver_utils.py

```diff
@@ -0,0 +1,123 @@
+        num_sections: int,
+        per_section_count: int,
+        per_section_constraints: Tuple[Tuple[str, str, int]],  # field, eq/leq/geq, val
+        overall_constraints: Tuple[Tuple[str, str, int]],  # field, eq/leq/geq, valS
+    ) -> Tuple[Tuple[Candidate]]:
+        """
+        Perform integer programming to solve constraints and maximize an objective function based on the total scores
+        applied to candidates. This function expects candidates that are the result of some recommendation pipeline
+        (i.e., candidates have scores, and problematic candidates have already been filtered out).
+
+        This will produce outputs assigning candidates to 'sections'. A section can be thought of as e.g. a day in
+        a meal plan, or a semester in a student's plan-of-study.
+
+        Currently assumes that (1) the objective function is always to maximize the total score of the final output,
+        (2) each candidate can only be a part of one section, (3) each section must have an exact number of
+        candidates assigned to it, and (4) the order of sections does not matter.
+
+        :param candidates: The candidates to choose for the optimization problem
+        :param num_sections: The number of sections to assign candidates to
+        :param per_section_count: The number of candidates that must be assigned into each section
+        :param per_section_constraints: Constraints to apply for each section. Currently using tuples of (str, str, int)
+        of the form (field_to_constrain, type_of_constraint (i.e., 'eq', 'leq', 'geq), value_to_constrain).
+        :param overall_constraints: Constraints for the overall solution. Currently using tuples of (str, str, int)
+        containing (field_to_constrain, type_of_constraint (i.e., 'eq', 'leq', 'geq), value_to_constrain).
+        :return: A tuple of tuples, containing candidates assigned to each section.
+        """
+        solver = pywraplp.Solver.CreateSolver("SCIP")
+
+        candidate_count = len(candidates)
+        solve_choice = {}
+        for i in range(candidate_count):
+            for j in range(num_sections):
+                solve_choice[i, j] = solver.IntVar(0, 1, "")
+
+        # each item is only assigned to one section
+        for i in range(candidate_count):
+            solver.Add(
+                solver.Sum([solve_choice[i, j] for j in range(num_sections)]) <= 1
+            )
+
+        # each section has exactly per_section_count items chosen
+        for j in range(num_sections):
+            solver.Add(
+                solver.Sum([solve_choice[i, j] for i in range(candidate_count)])
+                == per_section_count
+            )
+
+        # constraints to apply on each section, based on certain field names
+        for psc in per_section_constraints:
+            for j in range(num_sections):
+                ss = solver.Sum(
+                    [
+                        getattr(candidates[i].domain_object, psc[0])
```

Comment:
using getattr so that these constraints can just be specified using strings. not sure if there's a neater or more robust way of doing this in a generalized way.

Reply:
What you want is a Python equivalent of TypeScript's lookup types:
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-1.html
(keyof ...)

If you're using NamedTuple or @dataclass you can get the _fields, but there's no enumerated type of the field names.

Closest you could do is a string `enum` that throws an exception. But that introduces redundancy and isn't buying you much.

I'd stick with str for now. But do use a NamedTuple or @dataclass instead of Tuple[...] for the constraints. Then at least you can indicate .domain_object_attribute_name or similar.